### PR TITLE
Lifecycle tab

### PR DESCRIPTION
**Pull request by pr-watson**
### Description
This is the initial PR for creating the Lifecycle tab for the frontend. It will contain both a bar chart and table and display lifecycle data from our backend database. 

[RSPEED-344](https://issues.redhat.com/browse/RSPEED-344)

---

#### After:

---

### Checklist ☑️
- [ ] PR only fixes one issue or story 
- [ ] Change reviewed for extraneous code 
- [ ] UI best practices adhered to 
- [ ] Commits squashed and meaningfully named 
- [ ] All PR checks pass locally (build, lint, test, E2E)

##
- [ ] _(Optional) QE: Needs QE attention (OUIA changed, perceived impact to tests, no test coverage)_
- [ ] _(Optional) QE: Has been mentioned_
- [ ] _(Optional) UX: Needs UX attention (end user UX modified, missing designs)_
- [ ] _(Optional) UX: Has been mentioned_
##

### CODE REVIEW

File: src/Components/Lifecycle/lifecycle.tsx [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+      // Dispatch notif here
+      setIsLoading(false);
+      });
+  };
+
+  useEffect(() => {
+  const apiData : any  = fetchData()
+  setLifecycleChanges(apiData)
+
+  }, []);
```
Comment (hosekadam) [inline src/Components/Lifecycle/lifecycle.tsx:79]:
```suggestion
   const fetchData = async () => {
  setIsLoading(true);
  try {
      const data = await getLifecycleChanges();
      const upcomingChangesParagraphs = data || [];
      setLifecycleChanges(upcomingChangesParagraphs);
    } catch (error) {
      console.error("Error fetching lifecycle changes:", error);
    } finally {
      setIsLoading(false);
    }
  };

  useEffect(() => {
    fetchData();
  }, []);
```
Add the async and a small refactor of the fetching code - avoid the datarace in the code. Now it loads the data without getting `lifecycleData is not defined` error

File: src/constants.ts [hunk 2]
```diff
@@ -1,5 +1,6 @@
 export const DR_API = '/api/digital-roadmap/v1';
 export const DR_RELEASE_NOTES = '/release-notes';
+export const DR_LIFECYCLE = '/lifecycle/systems';
```
Comment (hosekadam) [inline src/constants.ts:3]:
```suggestion
export const DR_LIFECYCLE = '/lifecycle/systems/';
```
The api endopoint [has changed](https://github.com/RedHatInsights/digital-roadmap-backend/commit/e3fc115087714df42adc4a958fc387aa702a9c2a#diff-94f4d76d854beffce11d71669916f876ba7467a6a6a50abb6d13af2c09b20e3aL8) and we were getting redirect - bad data in `lifecycleData` which caused `lifecycleData.map is not a function`

File: src/Components/Lifecycle/lifecycle.tsx (hunk 1 above)
Comment (pr-watson) [reply src/Components/Lifecycle/lifecycle.tsx]: Thanks Adam I added your changes

File: src/Components/Lifecycle/lifecycle.scss [hunk 3]
```diff
@@ -0,0 +1,19 @@
+@import "~@redhat-cloud-services/frontend-components-utilities/styles/variables";
+
+.sample-component {
```
Comment (rebeccaalpert) [inline src/Components/Lifecycle/lifecycle.scss:3]: Are we using this or `non-relevant` anywhere?

File: src/Components/LifecycleChart/lifecycleChart.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    };
+
+    const getChart = (alert: any[], index: number) => {
+      const data: any[] = [];
+
+      alert?.map((datum: { packageType: string; }) => {
+        data.push({
+          ...datum,
+          x: alerts.length - index,
+          fill: datum.packageType === "Retired"
```
Comment (rebeccaalpert) [inline src/Components/LifecycleChart/lifecycleChart.tsx:43]: What I would suggest here is a call to a separate function with a switch statement based on datum.packageType, like this: 

```
const getPackageColor = () => {
     switch(datum.packageType) {
        case 'Retired':
          return "var(--pf-v5-global--danger-color--100)";
        // ... other cases go here
        case default:
          return;
     } 
}
```

The ternary statements work, but this would be cleaner and easier to modify if the list of colors changes/grows.

File: src/Components/LifecycleChart/lifecycleChart.tsx [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        <div style={{ height: "400px", width: "450px" }}>
+          <Chart
+            ariaDesc="Support timelines of packages and RHEL versions"
+            ariaTitle="Lifecycle bar chart"
+            containerComponent={
+              <ChartVoronoiContainer
+                labelComponent={
+                  <ChartTooltip constrainToVisibleArea labelComponent={<ChartLabel dx={-65} textAnchor="start" />} />
+                }
+                labels={({ datum }) => `Package: ${datum.x}\npackageType: ${datum.packageType}\nStart: ${formatDate(new Date(datum.y0), true)}\nEnd: ${formatDate(new Date(datum.y), true)}`}
```
Comment (rebeccaalpert) [inline src/Components/LifecycleChart/lifecycleChart.tsx:85]: It looks like these are getting cut off/not being formatted in a way that makes a lot of sense. I don't think it's worth holding this PR over, but I will create a Jira ticket to address this.

File: src/Components/LifecycleChart/lifecycleChart.tsx [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+            domainPadding={{ x: [20, 20], y: [40, 40] }}
+            legendData={[
+              { name: "Supported", symbol: { fill: "var(--pf-v5-global--success-color--100)" } },
+              { name: "Support ends within 6 months", symbol: { fill: "var(--pf-v5-global--warning-color--100)" } },
+              { name: "Retired", symbol: { fill: "var(--pf-v5-global--danger-color--100)" } },
+              { name: "Not installed", symbol: { fill: "var(--pf-v5-global--palette--blue-200)" } },
+              { name: "Upcoming release", symbol: { fill: "var(--pf-v5-global--palette--blue-100)" } }
+            ]}
+            legendPosition="bottom-left"
+            height={400}
```
Comment (rebeccaalpert) [inline src/Components/LifecycleChart/lifecycleChart.tsx:97]: This is fine for now as a first pass, but we'll want to make this full-width later. I have a Jira ticket for this, so no need to address right now.

File: src/Components/LifecycleTable/lifecycleTable.tsx [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    },
+    columnIndex
+  });
+
+  // In this example, we wrap all but the 1st column and make the 1st and 3rd columns sortable just to demonstrate.
+  return (
+    <Table aria-label="Sortable table" ouiaId="SortableTable">
+      <Thead>
+        <Tr>
+          <Th sort={getSortParams(0)}>{columnNames.name}</Th>
```
Comment (rebeccaalpert) [inline src/Components/LifecycleTable/lifecycleTable.tsx:99]: Sorting does not seem to work for Name (I am ok if we leave this and come back to it - let's just document it with a ticket).

File: src/Components/LifecycleTable/lifecycleTable.tsx [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+          <Th modifier="wrap" sort={getSortParams(1)}>
```
Comment (rebeccaalpert) [inline src/Components/LifecycleTable/lifecycleTable.tsx:100]: Trying to sort release is causing a bad bug - we should disable sort on this column if we can't handle this edge case or if we don't have good test data yet.

File: src/Components/LifecycleTable/lifecycleTable.tsx [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+          </Th>
+          <Th modifier="wrap" sort={getSortParams(4)}>
+            {columnNames.systems}
+          </Th>
+        </Tr>
+      </Thead>
+      <Tbody>
+        {sortedRepositories.map((repo: { name: string | number | boolean | React.ReactElement<any, string | React.JSXElementConstructor<any>> | Iterable<React.ReactNode> | React.ReactPortal | null | undefined; major: string | number | boolean | React.ReactElement<any, string | React.JSXElementConstructor<any>> | Iterable<React.ReactNode> | React.ReactPortal | null | undefined; minor: string | number | boolean | React.ReactElement<any, string | React.JSXElementConstructor<any>> | Iterable<React.ReactNode> | React.ReactPortal | null | undefined; release: string | number | boolean | React.ReactElement<any, string | React.JSXElementConstructor<any>> | Iterable<React.ReactNode> | React.ReactPortal | null | undefined; release_date: Moment.MomentInput; retirement_date: Moment.MomentInput; systems: string | number | boolean | React.ReactElement<any, string | React.JSXElementConstructor<any>> | Iterable<React.ReactNode> | React.ReactPortal | null | undefined; }, rowIndex: React.Key | null | undefined) => (
```
Comment (rebeccaalpert) [inline src/Components/LifecycleTable/lifecycleTable.tsx:115]: This is really big. Can we take this out and instead define a `type` or `interface` that we then mention here?

**Review by rebeccaalpert — CHANGES_REQUESTED** [review]
Looks good overall. I left a few inline comments. Some are things I'd like fixed if we can and others are just heads-up type things. Let me know if those aren't helpful - I'm happy to adjust.

Some other comments (not blocking, but want to note):

- We're going to want an error, loading states, empty state for pages (tracking elsewhere - no need to address here)
- We'd be well-served by removing Inline styles and migrating off of direct v5 selectors where possible to make it easier to upgrade to v6 later. A lot of components should accept a `className` prop. We may also want to consider using something like BEM for well-scoped class names: https://getbem.com/ 
- A lot of Red Hat repos are dropping/have dropped Moment, from what I recall. We are using localeCompare elsewhere. We may want to switch off Moment and to the native locale date functions at a later time. Not for this PR though - I will ticket it separately for after we have delivered core functionality.

File: src/Components/LifecycleTable/lifecycleTable.tsx (hunk 7 above)
Comment (pr-watson) [reply src/Components/LifecycleTable/lifecycleTable.tsx:99]: Yes currently Name only contains `RHEL`, i'll construct an array that combines RHEL with the minor and major versions

File: src/Components/LifecycleChart/lifecycleChart.tsx (hunk 5 above)
Comment (pr-watson) [reply src/Components/LifecycleChart/lifecycleChart.tsx:85]: Yes the chart has a few bugs to work out

File: src/Components/LifecycleChart/lifecycleChart.tsx (hunk 4 above)
Comment (pr-watson) [reply src/Components/LifecycleChart/lifecycleChart.tsx:43]: Sounds good i'll make this change

File: src/Components/LifecycleTable/lifecycleTable.tsx (hunk 8 above)
Comment (pr-watson) [reply src/Components/LifecycleTable/lifecycleTable.tsx:100]: What was the bug with release?

Comment (rebeccaalpert) [reply src/Components/LifecycleTable/lifecycleTable.tsx:100]: I'm seeing `TypeError
Cannot read properties of undefined (reading 'localeCompare')`


https://github.com/user-attachments/assets/d8fd2208-755f-44c0-ac3e-82163c9b0cbd

Comment (pr-watson) [reply src/Components/LifecycleTable/lifecycleTable.tsx:100]: Oh I see, we updated the mocks for the Release field to contain "Not applicable" now. After pulling in the latest changes for digital-roadmap-backend on the develop branch it should stop giving that error

**Review by rebeccaalpert — APPROVED** [review]
She's a beaut! Thank you so much for making all those changes.

**Review by hosekadam — APPROVED** [review]
